### PR TITLE
pre-commit: 2.20.0 -> 3.1.0

### PR DESCRIPTION
**Pull request by aaronjheng**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
aarch64-linux fails to build

**Comment by aaronjheng** [comment]
Done.

File: pkgs/tools/misc/pre-commit/default.nix [hunk 1]
```diff
@@ -82,7 +78,8 @@ buildPythonPackage rec {
   preCheck = ''
     export GIT_AUTHOR_NAME=test GIT_COMMITTER_NAME=test \
            GIT_AUTHOR_EMAIL=test@example.com GIT_COMMITTER_EMAIL=test@example.com \
-           VIRTUALENV_NO_DOWNLOAD=1 PRE_COMMIT_NO_CONCURRENCY=1 LANG=en_US.UTF-8
+           VIRTUALENV_NO_DOWNLOAD=1 PRE_COMMIT_NO_CONCURRENCY=1 LANG=en_US.UTF-8 \
+           RUSTFLAGS="-Ctarget-feature=-outline-atomics"
```
Comment (Jayman2000) [inline pkgs/tools/misc/pre-commit/default.nix:82]: It’s not clear to me why this was added. I would either add a comment or add an explanation to the body of the commit message.

**Review by Jayman2000 — CHANGES_REQUESTED** [review]
##### Reviewed points

- [x] package name fits guidelines
- [x] package version fits guidelines
- [x] package build on x86_64
- [x] executables tested on x86_64
- [x] all depending packages build

**Review by Jayman2000 — CHANGES_REQUESTED** [review]

File: pkgs/tools/misc/pre-commit/default.nix [hunk 2]
```diff
@@ -11,18 +11,18 @@
...
-  disabled = pythonOlder "3.6";
+  disabled = pythonOlder "3.8";
 
   src = fetchFromGitHub {
     owner = "pre-commit";
     repo = "pre-commit";
-    rev = "refs/tags/v${version}";
-    sha256 = "sha256-+JrnJz+wFbzVw9ysPX85DDE6suF3VU7gQZdp66x5TKY=";
+    rev = "v${version}";
+    hash = "sha256-+JrnJz+wFbzVw9ysPX85DDE6suF3VU7gQZdp66x5TKY=";
```
Comment (Jayman2000) [inline pkgs/tools/misc/pre-commit/default.nix:25]: I just ran `pre-commit --version` and it still said “2.20.0”. I think that this hash needs to be changed in order for a new source tarball to be used.

**Review by Jayman2000 — CHANGES_REQUESTED** [review]

**Comment by lovesegfault** [comment]
This does not build currently

File: pkgs/tools/misc/pre-commit/default.nix [hunk 3]
```diff
@@ -37,29 +41,23 @@ buildPythonPackage rec {
...
     pytest-xdist
     pytestCheckHook
     re-assert
-  ];
-
-  buildInputs = [
-    # Required for rust test on x86_64-darwin
-    libiconv
-  ];
+  ] ++ lib.optionals (stdenv.isDarwin && stdenv.isAarch64) [ libiconv ];
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/pre-commit/default.nix:60]: libiconv is null on linux so we could remove this check

**Review by SuperSandro2000 — APPROVED** [review]

**Review by Jayman2000 — APPROVED** [review]
##### Reviewed points

- [x] package name fits guidelines
- [x] package version fits guidelines
- [x] package build on x86_64
- [x] executables tested on x86_64
- [x] all depending packages build

**Pull request merged**